tools/fastqpreprocessing/src/fastq_common.cpp

```diff
@@ -426,22 +426,95 @@ void fillSamRecord(SamRecord* samRecord, FastQFile* fastQFileI1,
+                    break;
+                default:
+                    break;
+            }
+            cur_ind += length;
+        }
       }
       else if (strcmp(orientation.c_str(), "FIRST_BP_RC") == 0)
       {
-          barcode_seq = reverseComplement(sequence).substr(0, g_barcode_length);
-          reverse(quality_sequence.begin(), quality_sequence.end());
-          barcode_quality = quality_sequence.substr(0, g_barcode_length);
+        int cur_ind = 0;
+        for (auto [tag, length] : g_parsed_read_structure)
+        {
+            switch (tag)
+            {
+                case 'C':
+                    // For the barcode, append the reverse complement of the sequence
+                    barcode_seq += reverseComplement(sequence.substr(cur_ind, length));
+
+                     // Reverse the quality string to maintain correspondence
+                    reverse(quality_sequence.begin() + cur_ind, quality_sequence.begin() + cur_ind + length);
+                    barcode_quality += quality_sequence.substr(cur_ind, length);
+                    break;
+                case 'M':
+                    // For the UMI, append the reverse complement of the sequence
+                    umi_seq += reverseComplement(sequence.substr(cur_ind, length));
+
+                    // Reverse the quality string to maintain correspondence
+                    reverse(quality_sequence.begin() + cur_ind, quality_sequence.begin() + cur_ind + length);
+                    umi_quality += quality_sequence.substr(cur_ind, length);
+                    break;
+                default:
+                    break;
+            }
+            cur_ind += length;
+  
+        // old code below
+	      // barcode_seq = reverseComplement(sequence).substr(0, g_barcode_length);
+        // reverse(quality_sequence.begin(), quality_sequence.end());
+        // barcode_quality = quality_sequence.substr(0, g_barcode_length);
+	      // Print the barcode
+        // std::cout << "Barcode: " << barcode_seq << std::endl;
+	      // std::cout << "Sequence: " << sequence << std::endl;
+          }
       }
       else if (strcmp(orientation.c_str(), "LAST_BP_RC") == 0)
       {    
-          std::string reverse_complement = reverseComplement(sequence);
-          barcode_seq = reverse_complement.substr(reverse_complement.length() - g_barcode_length, reverse_complement.length());
-          
-          reverse(quality_sequence.begin(), quality_sequence.end());
-          barcode_quality = quality_sequence.substr(0, g_barcode_length);
+        std::string reverse_complement = reverseComplement(sequence);
+        barcode_seq = reverse_complement.substr(reverse_complement.length() - g_barcode_length, reverse_complement.length());
+    
+        reverse(quality_sequence.begin(), quality_sequence.end());
+        barcode_quality = quality_sequence.substr(0, g_barcode_length);
+
+        // Account for tags C and M
```

Comment:
Remove the section that accounts for tags in LAST_BP_RC section, which is only for ATAC so we do not need to account for it for paired-tag